### PR TITLE
ISSUE-27 make gringoft as third-party lib for other project

### PR DESCRIPTION
**Pull request by huyumars**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -248,12 +258,28 @@ add_subdirectory(src/app_demo)
 
 add_subdirectory(test)
 
+# expose to parent
+get_directory_property(hasParent PARENT_DIRECTORY)
+if(hasParent)
+    message(STATUS "Gringofts as a library")
+    set(GRINGOFTS_INCL
```
Comment (qiawu) [inline CMakeLists.txt:265]: what if Gringofts is placed under a folder ? We cannot use list here

File: src/infra/CMakeLists.txt [hunk 2]
```diff
@@ -1,45 +1,45 @@
...
+target_link_libraries(gringofts_infra_util gringofts_store_proto_library ${GRINGOFTS_LIBRARIES})
 
-add_library(infra STATIC
-        ${INFRA_SRC})
-add_dependencies(infra infra_raft store_proto_library)
-target_link_libraries(infra infra_raft store_proto_library infra_util)
+add_library(gringofts_infra STATIC
+        ${GRINGOFTS_INFRA_SRC})
+add_dependencies(gringofts_infra gringofts_infra_raft gringofts_store_proto_library)
+target_link_libraries(gringofts_infra gringofts_infra_raft gringofts_store_proto_library gringofts_infra_util)
```
Comment (qiawu) [inline src/infra/CMakeLists.txt:41]: `gringofts_store_proto_library` should be after `gringofts_infra_util`

**Comment by qiawu** [comment]
I created an issue to link this PR: #25

**Review by qiawu — APPROVED** [review]
LGTM

File: CMakeLists.txt [hunk 3]
```diff
@@ -192,12 +192,12 @@ include_directories(third_party/inih/cpp)
 
 # abseil-cpp
 set(CMAKE_CXX_STANDARD 17)
-add_subdirectory(third_party/abseil-cpp)
+add_subdirectory(third_party/abseil-cpp EXCLUDE_FROM_ALL)
```
Comment (jackyjia) [inline CMakeLists.txt:195]: why add this `EXCLUDE_FROM_ALL` flag?

File: CMakeLists.txt [hunk 4]
```diff
@@ -192,12 +192,12 @@ include_directories(third_party/inih/cpp)
...
 include_directories(third_party/abseil-cpp)
 set(ABSL_LIBS absl::strings absl::base)
 
 # prometheus
-add_subdirectory(third_party/prometheus-cpp)
+add_subdirectory(third_party/prometheus-cpp EXCLUDE_FROM_ALL)
 include_directories(third_party/prometheus-cpp/core/include)
 include_directories(third_party/prometheus-cpp/pull/include)
 set(MONITOR_LIBS prometheus-cpp::pull prometheus-cpp::core)
```
Comment (jackyjia) [inline CMakeLists.txt:203]: ditto

File: CMakeLists.txt [hunk 5]
```diff
@@ -258,12 +258,28 @@ add_subdirectory(src/app_demo)
 
 add_subdirectory(test)
 
+# expose to parent
+get_directory_property(hasParent PARENT_DIRECTORY)
+if(hasParent)
```
Comment (jackyjia) [inline CMakeLists.txt:263]: what changes are needed for Gringofts users?

File: CMakeLists.txt (hunk 3 above)
Comment (huyumars) [reply CMakeLists.txt:195]: https://cmake.org/cmake/help/v3.14/prop_tgt/EXCLUDE_FROM_ALL.html
only build what we want of absl by dependency

File: CMakeLists.txt (hunk 5 above)
Comment (huyumars) [reply CMakeLists.txt:263]: grinKV has an example, include below librarise set

**Review by jackyjia — APPROVED** [review]
LGTM. Thanks for the contribution!

**Pull request merged**